### ISSUE TITLE
spec: clarify empty volume kind

### ISSUE THREAD

**Issue by jonboulle** [appc/spec#183]
> kind (required) either "empty" or "host". "empty" fulfills a mount point by ensuring the path exists (writes go to container). "host" fulfills a mount point with a bind mount from a source.

"writes go to container" -> this is unclear as to whether this is shared between the apps in the container or not. I propose that empty means a (temporary)  directory that _is_ shared among the apps in the container. 

/cc @mpasternacki

**Comment by philips** [comment]
@jonboulle The volume is shared if it is mapped into two apps.

**Comment by cdaylward** [comment]
I presumed this would mean shared by all apps in the container (what I was aiming for). @jonboulle when you say temporary directory, do you mean a temporary directory (tmpnam()) on the host that is bind-mounted in?

**Comment by jonboulle** [comment]
@philips The point is that the wording ("writes go to container") is ambiguous; since each app has its own rootfs, if we just allowed it to be a writeable/empty directory it would not be shared. 

@cdaylward yes, that is essentially what I mean

**Comment by mpasternacki** [comment]
To clarify: a "temporary directory", as I understand it, would mean a directory whose lifetime is *at least* equal to lifetime of a container (i.e. survives container restart and host restart). Ideally, it could be "anonymously" passed from container to container, in a way similar to `--volumes-from` of Docker.

To clarify: Docker, if volume's source directory is not provided, creates a new unique directory and uses it as the volume source. When one creates a new container, a `--volumes-from=CONTAINER` can be provided, which reuses all host-bound and anonymous volumes from an existing container. If a container is removed, and it holds the last remaining reference to an anonymous volume, the volume is deleted (at least in theory). This allows, for example, for database upgrades without explicitly maintaining an index of data directories on the host. A simplified example:

```
docker run --name=postgres93 -d postgresql:9.3
# A /var/lib/postgresql volume is implicitly created, because the volume is declared in Dockerfile
docker stop postgresql93
docker run --name=postgresl94 --volumes-from=postgres93 postgresql:9.4
docker rm postgresql93
```

The idea of the created directory being "temporary" needs to be more precise: lifetime of the created directory needs to be known. I believe it should be at least equal to lifetime of the container, if not maintained separately and reference-count-garbage-collected as Docker does.

**Comment by jonboulle** [comment]
Reference: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/volumes.md#emptydir

**Comment by thockin** [comment]
I also flagged this as unclear in the spec.  I think you need to be clear that the volumes have at least the same lifetime as the pod, and equally accessible by any app in that pod

**Comment by jonboulle** [comment]
Let's also clarify the permitted fields for each volume
> Can we spec this whole block more as a conditional? "if kind is 'empty' ... fields; if kind is 'host' ... fields".

**Comment by jonboulle** [comment]
This was fixed by #449 but accidentally not closed.
